.circleci/config.yml

```diff
@@ -263,75 +246,90 @@ jobs:
         type: string
         default: "test/e2e"
     parallelism: << parameters.parallelism >>
-    environment:
-      GOPATH: /home/circleci/.go-kuma-go
     steps:
-      - install_build_tools
-      - checkout
-      - restore_cache:
-          keys:
-            # prefer the exact match
-            - vm-executor-go.mod-{{ .Branch }}-{{ checksum "go.sum" }}
-      - run:
-          name: "Download Go modules"
-          command: |
-            go mod download
-      # since execution of go commands might change contents of "go.sum", we have to save cache immediately
-      - save_cache:
-          key: vm-executor-go.mod-{{ .Branch }}-{{ checksum "go.sum" }}
-          paths:
-            - "/home/circleci/.go-kuma-go"
-      # Mount files from the upstream jobs
-      - attach_workspace:
-          at: build
       - run:
-          name: "Install all development tools"
+          name: dump params
           command: |
-            make dev/tools
-      - run:
-          name: "Setup Helm"
-          command: |
-            helm repo add kuma https://kumahq.github.io/charts
-      - when:
-          condition: << parameters.ipv6 >>
+            echo "Running with: \
+              ipv6:<< parameters.ipv6>> \
+              k8s:<< parameters.k8sVersion >> \
+              target:<< parameters.target >> \
+              parallelism:<< parameters.parallelism >> \
+            "
+
+            echo "some combinations may run nothing"
+      - unless:
+          # This works around circleci limitation by skipping tests for combinations that don't make sense
+          # See: https://discuss.circleci.com/t/matrix-exclude-entire-subset/43879
+          condition:
+            or:
+              - and:
+                  - or:
+                      - not: {matches: {pattern: "^release-.*", value: << pipeline.git.branch >>}}
+                      - not: {equal: [master, << pipeline.git.branch >>]}
+                  - or:
+                      - << parameters.ipv6 >>
+                      - equal: ["v1.19.16-k3s1", << parameters.k8sVersion >>]
+              - and:
+                # Ipv6 only works with kind
+                - << parameters.ipv6 >>
+                - not: {equal: [kind, << parameters.k8sVersion >>]}
+              - and:
+                # kind should only run with v6 except with universal
```

Comment:
do we need this `and` and the previous `and`?

Reply:
Yes it's an or. The first says `if ipv6 and not kind` don't run the other one says: `if not ipv6 and kind and not universal test`. Which is possible, I had a version with a `none` version but I found it even more complicated.